tests/node_suspension.py

```diff
@@ -0,0 +1,157 @@
+import os
+import getpass
+import time
+import logging
+import multiprocessing
+import shutil
+from random import seed
+import infra.ccf
+import infra.proc
+import infra.jsonrpc
+import infra.notification
+import infra.net
+import e2e_args
+from threading import Timer
+import random
+
+from loguru import logger as LOG
+
+# 256 is the number of messages PBFT keeps in memory before needing to replay the ledger
+# by setting 200 pure requests we will surpass 256 messages being passed throught the protocol
+TOTAL_REQUESTS = 200
+
+
+def timeout(node, suspend, election_timeout):
+    if suspend:
+        # We want to suspend the nodes' process so we need to initiate a new timer to wake it up eventually
+        if not node.suspend():
+            LOG.info("Node can not be suspended, probably has stopped running")
+            return
+        next_timeout = random.uniform(election_timeout, 3 * election_timeout)
+        LOG.info(f"New timer set for node {node.node_id} is {next_timeout} seconds")
+        t = Timer(next_timeout, timeout, args=[node, False, 0])
+        t.start()
+    else:
+        node.resume()
+
+
+def run(args):
+    hosts = ["localhost", "localhost", "localhost"]
+
+    with infra.notification.notification_server(args.notify_server) as notifications:
+        # Lua apps do not support notifications
+        # https://github.com/microsoft/CCF/issues/415
+        notifications_queue = (
+            notifications.get_queue() if args.package == "libloggingenc" else None
+        )
+
+        with infra.ccf.network(
+            hosts, args.build_dir, args.debug_nodes, args.perf_nodes, pdb=args.pdb
+        ) as network:
+            first_node, (backups) = network.start_and_join(args)
+
+            term_info = {}
+            long_msg = "X" * (2 ** 14)
+
+            # first timer determines after how many seconds each node will be suspended
+            t0 = random.uniform(1, 10)
+            LOG.info(f"Initial timer for primary is {t0} seconds...")
+            t1 = random.uniform(1, 10)
+            LOG.info(f"Initial timer for backup 1 is {t1} seconds...")
+            t2 = random.uniform(1, 10)
+            LOG.info(f"Initial timer for backup 2 is {t2} seconds...")
+            tm0 = Timer(
+                t0, timeout, args=[first_node, True, args.election_timeout / 1000],
+            )
+            tm0.start()
+            tm1 = Timer(
+                t1, timeout, args=[backups[0], True, args.election_timeout / 1000]
+            )
+            tm1.start()
+            tm2 = Timer(
+                t2, timeout, args=[backups[1], True, args.election_timeout / 1000]
+            )
+            tm2.start()
+
+            with first_node.node_client() as mc:
+                check_commit = infra.ccf.Checker(mc, notifications_queue)
+                check = infra.ccf.Checker()
+
+                LOG.info("Write messages to nodes using round robin")
+                with first_node.user_client(format="json") as c0:
```

Comment:
It would be nice to have this take a variable number of nodes/clients, to faciliate the test chaining stuff that we're about to do, but this is also good as a separate item of course. Also, and somewhat separately, you can `with A as a, B as b, C as c:` in a single scope, and if you have a variable number of them, and ExitStack is a good way to go: https://docs.python.org/3/library/contextlib.html#contextlib.ExitStack

Reply:
Thanks!